packages/monaco/src/browser/monaco-editor-model.ts

```diff
@@ -333,28 +337,52 @@ export class MonacoEditorModel implements ITextEditorModel, TextEditorDocument {
     }
 
     protected async fireWillSaveModel(reason: TextDocumentSaveReason, token: CancellationToken): Promise<void> {
-        const waitables: Thenable<monaco.editor.IIdentifiedSingleEditOperation[]>[] = [];
-
-        // The firing is synchronous.
-        // Every listener will register something for us to wait after.
-        this.onWillSaveModelEmitter.fire({
-            model: this, reason,
-            waitUntil: thenable => {
-                waitables.push(thenable);
+        type EditContributor = Thenable<monaco.editor.IIdentifiedSingleEditOperation[]>;
+        for (const listener of this.onWillSaveModelListeners) {
+            const waitables: EditContributor[] = [];
+            const { version } = this;
+
+            const event = {
+                model: this, reason,
+                waitUntil: (thenable: EditContributor) => {
+                    if (Object.isFrozen(waitables)) {
+                        throw new Error('waitUntil cannot be called asynchronously.');
+                    }
+                    waitables.push(thenable);
+                }
+            };
+
+            try {
+                listener(event);
+            } catch (err) {
+                console.error(err);
+                continue;
             }
-        });
 
-        // Wait for all listeners to resolve the edits.
-        return Promise.all(waitables).then(allOperations => {
+            // Asynchronous calls to `waitUntil` should fail.
+            Object.freeze(waitables);
+
+            // Wait for all promises and apply edits.
+            const edits = await Promise.all(waitables).then(allOperations =>
+                ([] as monaco.editor.IIdentifiedSingleEditOperation[]).concat(...allOperations)
+            );
             if (token.isCancellationRequested) {
                 return;
             }
-            for (const operations of allOperations) {
-                this.applyEdits(operations, {
+
+            // Only apply edits if the document did not change.
+            if (version !== this.version) {
```

Comment:
I don't think we need this check, it should be already ensured by `token.isCancellationRequested`, the operation is cancelled whenever there are changes to the document.

Reply:
Apparently I manage to get down to this point, despite the cancellation check before it (I can see an error thrown by this content change check)

Reply:
ok